### ISSUE TITLE
Validity of the WIDE-Part

### ISSUE THREAD

**Issue by aendk** [jrzaurin/pytorch-widedeep#112]
Hey!

I have been using parts of your great repository, but there is one thing I am confused about:

The [original Cross-Product Transformation](https://datascience.stackexchange.com/questions/57435/how-is-the-cross-product-transformation-defined-for-binary-features) is defined as a _product_, wherein `c` selects/deselects the relevant embeddings which are then multiplied together.
In your [WIDE](https://github.com/jrzaurin/pytorch-widedeep/blob/master/pytorch_widedeep/models/tabular/linear/wide.py#L65), the outputs of the underlying embedding layer are _added_ together in the forward pass.

Does this difference matter or am I overlooking something?

**Comment by jrzaurin** [comment]
Hey @aendk 

Thanks for your words, and let me see if I can explain our implementation of the wide component. Let me start by saying that strictly speaking, the wide component does not use Embeddings, we implement is as Embeddings, but is a linear model. In other words, is literally:   

$$ y = \textbf{W}^T \textbf{x} + b $$

Where the none linearities are captured via crossed products. If you wanted to see how a linear layer (nn.Linear) is equivalent to our implementation using an Embedding Layer of dim 1 (but this is much more efficient) please, look [here](https://github.com/jrzaurin/pytorch-widedeep/blob/master/examples/notebooks/02_model_components.ipynb). 

Now, a cross product is defined in the paper as: "_a boolean variable that is 1 if the i-th feature is part of the k-th transformation_ $\Phi_{k}$ , _and 0 otherwise. For binary features, a cross-product transformation (e.g., “AND(gender=female, language=en)”) is 1 if and only if the constituent features (“gender=female” and “language=en”) are all 1, and 0 otherwise_". We implement this at "preprocessing time". Imagine we have a dataset like this:

```python
>>> import pandas as pd
>>> df = pd.DataFrame({'color': ['r', 'b', 'g'], 'size': ['s', 'n', 'l']})
>>> df
  color size
0     r    s
1     b    n
2     g    l
```

For a given model, we decided that we want to use, as the wide component, the `color` column and the cross product of the `color` and `size` columns. To do such a thing we would do: 

```python
>>> from pytorch_widedeep.preprocessing import WidePreprocessor
>>> wide_cols = ['color']
>>> crossed_cols = [('color', 'size')]
>>> wide_preprocessor = WidePreprocessor(wide_cols=wide_cols, crossed_cols=crossed_cols)
>>> X_wide = wide_preprocessor.fit_transform(df)
>>> wide_preprocessor.encoding_dict
{'color_r': 1, 'color_b': 2, 'color_g': 3, 'color_size_r-s': 4, 'color_size_b-n': 5, 'color_size_g-l': 6}
>>> wide_preprocessor.inverse_transform(X_wide)
  color color_size
0     r        r-s
1     b        b-n
2     g        g-l
```

As you can see, we have fit a preprocessor that has a series of attributes, amongst them an encoding dictionary that encodes the 6 possible different values that all features (`color` and `color_size`) could take. We have implemented the crossed product as effectively a new column.

In fact the fitted preprocessor has a method `inverse_transform` that will return an "inverted transformed" dataframe relative to the column/features that were passed to the `WidePreprocessor` preprocessor. 

So far so good (I hope). Now onto the model. The wide component is, as I mentioned, a linear model that is eventually added to the deep component (see equation 3 in the [original paper](https://arxiv.org/pdf/1606.07792.pdf). We implement this as an Embedding Layer with dim 1. Let's have a look: 

```python
>>> import torch
>>> from pytorch_widedeep.models import Wide
>>> Xt_Wide = torch.from_numpy(X_wide)
>>> wide = Wide(input_dim=Xt_Wide.unique().size(0), pred_dim=1)
>>> wide
Wide(
  (wide_linear): Embedding(7, 1, padding_idx=0)
)
>>> out = wide(Xt_Wide)
>>> out
tensor([[0.8094],
        [0.2389],
        [0.4096]], grad_fn=<AddBackward0>)
```

As you can see our wide model is an embedding layer of 7 weights (the 6 possible values and 0, which is reserved for "unseen" values at inference time) 

So, to make it even more specific, our implementation for the 1st row of the transformed dataset is effectively this:

$$
out[0] = W_{r} \times 1 +  W_{b} \times 0 +  W_{g} \times 0 +  W_{r-s} \times 1 +  W_{b-n} \times 0 + W_{g-l} \times 0 + bias = W_{r} + W_{r-s} + bias
$$

Which is effectively the implementation in the paper. The this wide output is added to the deep output and passed through the activation function (see again their equation 3) 

Let me know if this solves your question or you need more info/help

**Comment by aendk** [comment]
Thank you so much @jrzaurin, this really helped me/our team; I've overlooked the required preprocessing and expected the `Wide`-layer  to "just work".

I'll dig deeper into your library and your `WidePreprocessor`

**Comment by jrzaurin** [comment]
If you have any questions/suggestion let me know!
